modules/activiti-engine/src/main/java/org/activiti/engine/impl/persistence/entity/SignalEventSessionManager.java

```diff
@@ -0,0 +1,122 @@
+package org.activiti.engine.impl.persistence.entity;
+
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.activiti.engine.history.HistoricActivityInstance;
+import org.activiti.engine.impl.HistoricActivityInstanceQueryImpl;
+import org.activiti.engine.impl.bpmn.behavior.IntermediateThrowSignalEventActivityBehavior;
+import org.activiti.engine.impl.context.Context;
+import org.activiti.engine.impl.interceptor.CommandContext;
+import org.activiti.engine.impl.persistence.AbstractManager;
+import org.activiti.engine.impl.pvm.delegate.ActivityBehavior;
+import org.activiti.engine.impl.pvm.delegate.ActivityExecution;
+
+/**
+ * Keep track of fired events for process instance execution 
+ * in the current command context session and historic activity instance 
+ *
+ */
+public class SignalEventSessionManager extends AbstractManager {
+
+  /** collect fired events in process instance scope in the current command context session */
+  protected Map<String,Set<String>> firedSignalEventsSesssionRegistry = new HashMap<String, Set<String>>();

```

Comment:
The state is managed in memory, but what happens when each execution executes separately in a cluster environment?